lesson-04-memory-management/mm/xxx.c

```diff
@@ -5,25 +5,136 @@
+	char buf[BUFFER_SIZE];
+};
+
+static struct kmem_cache* pmem_cache = NULL;
+
+const char* get_mem_type(int type)
+{
+	switch (type) {
+		case MEM_TYPE_KMALLOC:
+			return "MEM_TYPE_KMALLOC";
+		case MEM_TYPE_KCACHE:
+			return "MEM_TYPE_KCACHE";
+		case MEM_TYPE_KVMALLOC:
+			return "MEM_TYPE_KVMALLOC";
+	}
+	return "UNKNOWN";
+}
+
+static void* mem_alloc(int type, size_t size)
+{
+	void* buf = NULL;
+	switch (type) {
+		case MEM_TYPE_KMALLOC:
+			buf = kmalloc(size, GFP_KERNEL);
+			break;
+		case MEM_TYPE_KCACHE:
+			buf = kmem_cache_alloc( pmem_cache, GFP_KERNEL );
+			break;
+		case MEM_TYPE_KVMALLOC:
+			buf = vmalloc(size);
+			break;
+	}
+
+	if (buf) buf_size = size;
+
+	pr_info("[%s:%s] {%s} %zu bytes --> %p\n", THIS_MODULE->name, __FUNCTION__, get_mem_type(type), size, buf);
+	return buf;
+}
+
+static void mem_free(int type, void* buf)
+{
+	if (!buf) {
+		pr_warning("[%s:%s] Wrong parameter %p\n", THIS_MODULE->name, __FUNCTION__, buf);
+		return;
+	}
+	switch (type) {
+		case MEM_TYPE_KMALLOC:
+			kfree(buf);
+			break;
+		case MEM_TYPE_KCACHE:
+			kmem_cache_free( pmem_cache, buf );
+			buf = kmem_cache_alloc( pmem_cache, GFP_KERNEL );
+			break;
+		case MEM_TYPE_KVMALLOC:
+			vfree(buf);
+			break;
+	}
+	pr_info("[%s:%s] {%s} bytes --> %p\n", THIS_MODULE->name, __FUNCTION__, get_mem_type(type), buf);
+}
+
+void init_kcache(void) 
+{
+	if (mem_type != MEM_TYPE_KCACHE) return;
+/*
+ * 140 * Please use this macro to create slab caches. Simply specify the
+ * 141 * name of the structure and maybe some flags that are listed above.
+ * 142 *
+ * 143 * The alignment of the struct determines object alignment. If you
+ * 144 * f.e. add ____cacheline_aligned_in_smp to the struct declaration
+ * 145 * then the objects will be properly aligned in SMP configurations.
+ * 146
+ * #define KMEM_CACHE(__struct, __flags) kmem_cache_create(#__struct,\
+ * 148  sizeof(struct __struct), __alignof__(struct __struct),\
+ * 149  (__flags), NULL)
+ */
+	pmem_cache = KMEM_CACHE(KCACHE_STRUCT, 0);
+	pr_info("[%s:%s] pmem_cache: %p\n", THIS_MODULE->name, __FUNCTION__, pmem_cache);
+}
+
+void deinit_kcache(void)
+{
+	if (mem_type != MEM_TYPE_KCACHE) {
+		mem_free(mem_type, pbuf_msg);
+		return;
+	}
+	kmem_cache_destroy( pmem_cache );
+}
 
 
 static ssize_t xxx_show(struct class *class, struct class_attribute *attr,
 						char *buf)
 {
-	strcpy(buf, buf_msg);
-	pr_info("read %ld\n", (long)strlen(buf));
-	return strlen(buf);
+	if (!pbuf_msg) {
+		printk( "[%s:%s] nothing to read\n", THIS_MODULE->name, __FUNCTION__);
+		return 0; 
+	}
+	strcpy( buf, pbuf_msg );
+	pr_info( "[%s:%s] read %ld\n", THIS_MODULE->name, __FUNCTION__, (long)strlen( buf ) );
+	return strlen( buf );
 }
 
 static ssize_t xxx_store(struct class *class, struct class_attribute *attr,
 						 const char *buf, size_t count)
 {
-	pr_info("write %ld\n", (long)count);
-	strncpy(buf_msg, buf, count);
-	buf_msg[count] = '\0';
+	if (!pbuf_msg) {
+		pbuf_msg = mem_alloc(mem_type, BUFFER_SIZE);
```

Comment:
The idea was to allocate required amount of memory dynamically.
Not to do that once (even if on first write), but be able to adjust buffer size at least for bigger `count`.